File: my-folder/0146-lru-cache/solution.java
class LRUCache {
    
    // head - most recently used
    // tail - least recently used
    class Node{
        int key;
        int value;
        Node pre;
        Node next;
    }
    int capacity;
    HashMap<Integer, Node> cache = new HashMap<>();
    Node head;
    Node tail;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        head = new Node();
        tail = new Node();
        head.next = tail;
        tail.pre = head;
    }
    
    public int get(int key) {
        if(cache.containsKey(key)){
            Node node = cache.get(key);
            // move to head
            moveToHead(node);
            return node.value;
        }
        return -1;
        
    }
    
    private void moveToHead(Node node){
        // remove
        remove(node);
        // add to head
        addToHead(node);
    }
    
    private void remove(Node node){
        node.pre.next = node.next;
        node.next.pre = node.pre;
    }
    
    private void addToHead(Node node){
        node.pre = head;
        node.next = head.next;
        head.next.pre = node;
        
        head.next = node;
    }
    
    public void put(int key, int value) {
        if(cache.containsKey(key)){
            // update value;
            Node node = cache.get(key);
            node.value = value;
            // move to head;
            moveToHead(node);
        }else{
            Node newNode = new Node();
            newNode.key = key;
            newNode.value = value;
            
            cache.put(key, newNode);
            addToHead(newNode);
            
            // remove from tail
            if(cache.size()>capacity){
                Node toRemove = tail.pre;
                remove(toRemove);
                cache.remove(toRemove.key);
            }
            
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: my-folder/0146-lru-cache/solution.py
class LRUCache:

    def __init__(self, capacity: int):
        self.cap = capacity
        self.cache = {}
        self.used = []
    def get(self, key: int) -> int:
        if key in self.cache:
            self.used.remove(key)
            self.used.append(key)
            return self.cache[key]
        else:
            return -1

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.used.remove(key)
            self.used.append(key)
            self.cache[key] = value
        else:
            while len(self.cache) >= self.cap:
                delete_key = self.used.pop(0)
                del self.cache[delete_key]
            self.used.append(key)
            self.cache[key] = value
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
